File: diet/calculator.py
"""Расчёт нормы калорий и БЖУ.

Доступны две формулы базового обмена (BMR):
    - WHO:  уравнения Скофилда, официальный стандарт ФАО/ВОЗ/ООН.
            Зависит от пола и возрастной группы, использует вес (кг) и рост (м).
    - mifflin: формула Миффлина-Сан Жеора. Современная альтернатива,
            чаще точнее для современных людей (Скофилд иногда завышает).

Расчёт БЖУ зависит от ДВУХ параметров:
    - condition:  нозологическая группа (диабет, ХБП, ожирение, ССЗ, здоровый);
    - life_stage: стадия жизни (спортсмен, пожилой, беременная, кормящая, ...).

Цепочка расчёта:
    BMR  → TDEE (= BMR × коэффициент активности)
        → целевые ккал (TDEE + поправка на цель + прибавка стадии жизни,
                         не ниже минимума)
        → БЖУ (с приоритетом CKD по белку — см. _calc_protein)
"""

from __future__ import annotations

from dataclasses import dataclass, field

from .profile import (
    ACTIVITY_LEVELS,
    FEMALE,
    GOALS,
    MALE,
    MIN_KCAL,
    UserProfile,
)
from .conditions import (
    CONDITIONS,
    PROTEIN_GKG as COND_PROTEIN_GKG,
    PROTEIN_PERCENT,
    VALID_CONDITIONS,
    get_condition,
)
from .life_stages import (
    LIFE_STAGES,
    PROTEIN_ADD_G,
    PROTEIN_GKG as STAGE_PROTEIN_GKG,
    VALID_LIFE_STAGES,
    get_life_stage,
)

# Калорийность 1 грамма макронутриента, ккал/г.
KCAL_PER_GRAM = {"protein": 4.0, "fat": 9.0, "carbs": 4.0}

# Жиры как доля энергии во всех группах (≤30%, ВОЗ/AHA).
FAT_SHARE = 0.30


# --- Уравнения Скофилда (ФАО/ВОЗ/ООН, 1985), ккал; W=кг, H=метры ----------------
# Источник: FAO "Energy and protein requirements", Annex 1.
# (коэффициенты взяты прямо из таблицы, проверены по примеру в Annex 1.B)
_WHO_BMR = {
    MALE: {
        (10, 18): lambda W, H: 16.6 * W + 77.0 * H + 572.0,
        (18, 30): lambda W, H: 15.4 * W - 27.0 * H + 717.0,
        (30, 60): lambda W, H: 11.3 * W + 16.0 * H + 901.0,
        (60, 200): lambda W, H: 8.8 * W + 1128.0 * H - 1071.0,
    },
    FEMALE: {
        (10, 18): lambda W, H: 7.4 * W + 482.0 * H + 217.0,
        (18, 30): lambda W, H: 13.3 * W + 334.0 * H + 35.0,
        (30, 60): lambda W, H: 8.7 * W - 25.0 * H + 865.0,
        (60, 200): lambda W, H: 9.2 * W + 637.0 * H - 302.0,
    },
}

VALID_FORMULAS = ("who", "mifflin")


@dataclass
class NutritionResult:
    """Итог расчёта — норма калорий и БЖУ."""

    formula: str          # формула, по которой считали итог (who / mifflin)
    condition: str        # нозологическая группа
    condition_label: str  # читаемое название группы
    life_stage: str       # стадия жизни
    life_stage_label: str # читаемое название стадии
    notes: str | None     # текстовые рекомендации (склейка condition + life_stage)
    warnings: list[str] = field(default_factory=list)  # тревожные комбинации

    bmr_mifflin: float = 0.0
    bmr_who: float = 0.0

    bmr: float = 0.0
    tdee: float = 0.0
    tdee_mifflin: float = 0.0
    tdee_who: float = 0.0

    activity_label: str = ""
    goal_label: str = ""
    target_kcal: float = 0.0

    protein_g: float = 0.0
    fat_g: float = 0.0
    carbs_g: float = 0.0

    @property
    def protein_kcal(self) -> float:
        return round(self.protein_g * KCAL_PER_GRAM["protein"])

    @property
    def fat_kcal(self) -> float:
        return round(self.fat_g * KCAL_PER_GRAM["fat"])

    @property
    def carbs_kcal(self) -> float:
        return round(self.carbs_g * KCAL_PER_GRAM["carbs"])

    def summary(self) -> dict:
        """Плоский словарь для вывода в таблицу pandas."""
        return {
            "Группа здоровья": self.condition_label,
            "Стадия жизни": self.life_stage_label,
            "Формула BMR": self.formula,
            "BMR по ВОЗ, ккал": round(self.bmr_who),
            "BMR по Миффлину, ккал": round(self.bmr_mifflin),
            "BMR используемый, ккал": round(self.bmr),
            "Расход за день (TDEE), ккал": round(self.tdee),
            "Уровень активности": self.activity_label,
            "Цель": self.goal_label,
            "Целевые калории, ккал/день": round(self.target_kcal),
            "Белки, г": round(self.protein_g),
            "Жиры, г": round(self.fat_g),
            "Углеводы, г": round(self.carbs_g),
        }

    def comparison(self) -> dict:
        """Сравнение двух формул BMR/TDEE (для отдельной таблицы)."""
        return {
            "BMR, ккал": {
                "ВОЗ (Скофилд)": round(self.bmr_who),
                "Миффлин-Сан Жеор": round(self.bmr_mifflin),
            },
            "Расход за день (TDEE), ккал": {
                "ВОЗ (Скофилд)": round(self.tdee_who),
                "Миффлин-Сан Жеор": round(self.tdee_mifflin),
            },
        }


def calc_bmr_mifflin(profile: UserProfile) -> float:
    """BMR по формуле Миффлина-Сан Жеора."""
    base = 10 * profile.weight + 6.25 * profile.height - 5 * profile.age
    return base + 5 if profile.sex == MALE else base - 161


def _who_equation(sex: str, age: int):
    """Выбирает уравнение Скофилда по полу и возрастной группе."""
    for (lo, hi), eq in _WHO_BMR[sex].items():
        if lo <= age < hi:
            return eq
    raise ValueError(f"Нет уравнения ВОЗ для возраста {age}")


def calc_bmr_who(profile: UserProfile) -> float:
    """BMR по официальным уравнениям ФАО/ВОЗ/ООН (Скофилд).

    Рост в формулах берётся в метрах.
    """
    eq = _who_equation(profile.sex, profile.age)
    return eq(profile.weight, profile.height / 100.0)


def calc_tdee(bmr: float, activity_key: str) -> float:
    """Суточный расход энергии = BMR × коэффициент активности."""
    return bmr * ACTIVITY_LEVELS[activity_key]["factor"]


def calc_target_kcal(tdee: float, profile: UserProfile, kcal_add: float = 0) -> float:
    """Целевая калорийность: TDEE + поправка на цель + прибавка стадии жизни,
    но не ниже безопасного минимума."""
    delta = GOALS[profile.goal]["kcal_delta"]
    target = tdee + delta + kcal_add
    return max(target, MIN_KCAL[profile.sex])


def _calc_protein(profile: UserProfile, condition_key: str, life_stage_key: str,
                  condition_kcal_basis: float) -> tuple[float, str | None]:
    """Считает белок (г) с приоритетом разрешения конфликтов.

    Возвращает (граммы белка, основание решения — для предупреждения, если есть).

    ПРИОРИТЕТ:
        1. CKD всегда выигрывает по белку (0.8 г/кг) — защита почек.
           Если при этом выбрана стадия жизни с повышенным белком —
           возвращаем предупреждение.
        2. Иначе: стадия жизни (если не default) перекрывает condition,
           т.к. это активная потребность живого человека.
        3. Иначе: condition (его gkg или percent).
    """
    cond_cfg = get_condition(condition_key)
    stage_cfg = get_life_stage(life_stage_key)
    warning = None

    is_ckd = (condition_key == "ckd")

    # 1) CKD имеет абсолютный приоритет по белку
    if is_ckd:
        protein_g = cond_cfg["protein_gkg"] * profile.weight
        if stage_cfg["protein_mode"] in (STAGE_PROTEIN_GKG, PROTEIN_ADD_G):
            warning = (
                f"При ХБП белок ограничен до {cond_cfg['protein_gkg']} г/кг "
                f"вне зависимости от стадии жизни ({stage_cfg['label']}). "
                f"Стадию/диализ уточняет нефролог."
            )
        return protein_g, warning

    # 2) Стадия жизни
    mode = stage_cfg["protein_mode"]
    if mode == STAGE_PROTEIN_GKG:
        return stage_cfg["protein_gkg"] * profile.weight, None
    if mode == PROTEIN_ADD_G:
        # прибавка к БАЗЕ — базе служит белок по condition (percent/gkg)
        base = _protein_from_condition(profile, cond_cfg, condition_kcal_basis)
        return base + stage_cfg["protein_add_g"], None

    # 3) Condition
    return _protein_from_condition(profile, cond_cfg, condition_kcal_basis), None


def _protein_from_condition(profile: UserProfile, cond_cfg: dict, kcal_basis: float) -> float:
    """Белок по настройкам condition: либо г/кг, либо % от энергии."""
    if cond_cfg["protein_mode"] == COND_PROTEIN_GKG:
        return cond_cfg["protein_gkg"] * profile.weight
    # PROTEIN_PERCENT
    return (kcal_basis * cond_cfg["protein_pct"]) / KCAL_PER_GRAM["protein"]


def _build_warnings(profile: UserProfile, condition_key: str, life_stage_key: str,
                    goal: str, protein_warning: str | None) -> list[str]:
    """Собирает список предупреждений о рискованных комбинациях."""
    warnings = []
    if protein_warning:
        warnings.append(protein_warning)
    if life_stage_key == "pregnant" and goal == "lose":
        warnings.append(
            "Похудение во время беременности не рекомендуется — нужна "
            "консультация врача. Расчёт дан, но применять с осторожностью."
        )
    if life_stage_key in ("pregnant", "lactating") and condition_key in ("diabetes_t2", "ckd"):
        warnings.append(
            f"{get_life_stage(life_stage_key)['label']} + "
            f"{get_condition(condition_key)['label']} — сложная комбинация, "
            f"требует ведения под контролем врача."
        )
    return warnings


def calc_macros(profile: UserProfile, target_kcal: float, condition_key: str,
                life_stage_key: str) -> tuple[float, float, float, str | None]:
    """Распределение БЖУ в граммах.

    Белок — по приоритету (см. _calc_protein).
    Жиры — FAT_SHARE от энергии. Углеводы — остаток.
    Возвращает (белок, жиры, углеводы, protein_warning).
    """
    protein_g, protein_warning = _calc_protein(
        profile, condition_key, life_stage_key, target_kcal
    )
    protein_kcal = protein_g * KCAL_PER_GRAM["protein"]

    fat_g = (target_kcal * FAT_SHARE) / KCAL_PER_GRAM["fat"]
    fat_kcal = fat_g * KCAL_PER_GRAM["fat"]

    carbs_kcal = target_kcal - protein_kcal - fat_kcal
    carbs_g = carbs_kcal / KCAL_PER_GRAM["carbs"]

    return protein_g, fat_g, carbs_g, protein_warning


def _join_notes(cond_cfg: dict, stage_cfg: dict) -> str | None:
    """Склеивает заметки condition и life_stage (если есть)."""
    parts = []
    if cond_cfg.get("notes"):
        parts.append(cond_cfg["notes"])
    if stage_cfg.get("notes"):
        parts.append(stage_cfg["notes"])
    return "\n\n".join(parts) if parts else None


def calculate(profile: UserProfile, formula: str = "who",
              condition: str = "healthy", life_stage: str = "default") -> NutritionResult:
    """Полный расчёт нормы калорий и БЖУ для профиля.

    formula:    "who" (по умолчанию) или "mifflin".
    condition:  нозологическая группа (healthy / diabetes_t2 / obesity / ckd / cvd).
    life_stage: стадия жизни (default / athlete_endurance / athlete_strength /
                older_adult / pregnant / lactating).
    """
    if formula not in VALID_FORMULAS:
        raise ValueError(f"formula должна быть одной из {VALID_FORMULAS}")
    profile.validate()
    cond_cfg = get_condition(condition)
    stage_cfg = get_life_stage(life_stage)

    bmr_mifflin = calc_bmr_mifflin(profile)
    bmr_who = calc_bmr_who(profile)

    tdee_mifflin = calc_tdee(bmr_mifflin, profile.activity)
    tdee_who = calc_tdee(bmr_who, profile.activity)

    if formula == "who":
        bmr, tdee = bmr_who, tdee_who
    else:
        bmr, tdee = bmr_mifflin, tdee_mifflin

    kcal_add = stage_cfg.get("kcal_add", 0)
    target_kcal = calc_target_kcal(tdee, profile, kcal_add=kcal_add)
    protein_g, fat_g, carbs_g, protein_warning = calc_macros(
        profile, target_kcal, condition, life_stage
    )

    warnings = _build_warnings(
        profile, condition, life_stage, profile.goal, protein_warning
    )

    return NutritionResult(
        formula=formula,
        condition=condition,
        condition_label=cond_cfg["label"],
        life_stage=life_stage,
        life_stage_label=stage_cfg["label"],
        notes=_join_notes(cond_cfg, stage_cfg),
        warnings=warnings,
        bmr_mifflin=bmr_mifflin,
        bmr_who=bmr_who,
        bmr=bmr,
        tdee=tdee,
        tdee_mifflin=tdee_mifflin,
        tdee_who=tdee_who,
        activity_label=ACTIVITY_LEVELS[profile.activity]["label"],
        goal_label=GOALS[profile.goal]["label"],
        target_kcal=target_kcal,
        protein_g=protein_g,
        fat_g=fat_g,
        carbs_g=carbs_g,
    )

File: web/frontend/src/lib/calculator.ts
/**
 * Расчёт нормы калорий, БЖУ и клетчатки.
 *
 * Портировано 1:1 с diet/calculator.py (Python). Вся арифметика идентична:
 * формулы ВОЗ (Скофилд) и Миффлина-Сан Жеора, приоритеты белка (CKD > стадия
 * жизни > condition), клетчатка по норме группы (не ниже 25 г/день).
 *
 * Цепочка:
 *   BMR → TDEE (× factor активности) → целевые ккал (+Δ цели, ≥ минимума)
 *       → БЖУ (белок по приоритету, жиры 30%, углеводы остаток)
 *       → клетчатка (от калорий, ≥ 25 г)
 */

// --- Константы -------------------------------------------------------------

const KCAL_PER_GRAM = { protein: 4.0, fat: 9.0, carbs: 4.0 } as const;
const FAT_SHARE = 0.30; // жиры как доля энергии (≤30%, ВОЗ/AHA)
const MIN_FIBER_G = 25.0; // минимум клетчатки для взрослых (DRI)
const MIN_KCAL = { male: 1500, female: 1200, other: 1200 } as const;

// --- Справочники (перенесены из profile.py / conditions.py / life_stages.py) -

export const ACTIVITY_LEVELS: Record<string, { label: string; factor: number }> = {
  // базовые (используются в онбординге)
  sedentary: { label: 'Сидячий', factor: 1.20 },
  light: { label: 'Лёгкая', factor: 1.375 },
  moderate: { label: 'Умеренная', factor: 1.55 },
  high: { label: 'Высокая', factor: 1.725 },
  very_high: { label: 'Очень высокая', factor: 1.90 },
  // онбординг использует 'very'/'extra' — маппим на ближайшие
  very: { label: 'Высокая', factor: 1.725 },
  extra: { label: 'Очень высокая', factor: 1.90 },
  // детальные пресеты
  light_cardio: { label: 'Лёгкая: кардио', factor: 1.375 },
  light_strength: { label: 'Лёгкая: силовые', factor: 1.35 },
  light_mixed: { label: 'Лёгкая: смесь', factor: 1.375 },
  moderate_cardio: { label: 'Умеренная: кардио', factor: 1.60 },
  moderate_strength: { label: 'Умеренная: силовые', factor: 1.50 },
  moderate_mixed: { label: 'Умеренная: смесь', factor: 1.55 },
  high_cardio: { label: 'Высокая: кардио', factor: 1.75 },
  high_strength: { label: 'Высокая: силовые', factor: 1.65 },
  high_mixed: { label: 'Высокая: смесь', factor: 1.725 },
};

export const GOALS: Record<string, { label: string; kcal_delta: number }> = {
  maintain: { label: 'Поддержание веса', kcal_delta: 0 },
  lose: { label: 'Снижение веса', kcal_delta: -500 },
  gain: { label: 'Набор веса', kcal_delta: 300 },
};

// Маппинг ключей онбординга → ключи калькулятора.
const CONDITION_MAP: Record<string, string> = {
  healthy: 'healthy',
  diabetes: 'diabetes_t2',
  obesity: 'obesity',
  hypertension: 'cvd',
  kidney: 'ckd',
  // онбординг может прислать и прямые ключи калькулятора
  diabetes_t2: 'diabetes_t2',
  cvd: 'cvd',
  ckd: 'ckd',
  // неизвестные → здоровый
};

type ProteinMode = 'percent' | 'gkg' | 'add_g' | null;

interface ConditionCfg {
  label: string;
  protein_mode: 'percent' | 'gkg';
  protein_pct?: number; // при percent
  protein_gkg?: number; // при gkg
  fiber_per_kcal: number;
}

const CONDITIONS: Record<string, ConditionCfg> = {
  healthy: { label: 'Здоровый', protein_mode: 'percent', protein_pct: 0.15, fiber_per_kcal: 0.014 },
  diabetes_t2: { label: 'Сахарный диабет 2 типа', protein_mode: 'percent', protein_pct: 0.18, fiber_per_kcal: 0.015 },
  obesity: { label: 'Ожирение / снижение веса', protein_mode: 'gkg', protein_gkg: 1.3, fiber_per_kcal: 0.015 },
  ckd: { label: 'Хроническая болезнь почек', protein_mode: 'gkg', protein_gkg: 0.8, fiber_per_kcal: 0.012 },
  cvd: { label: 'ССЗ / гипертония', protein_mode: 'percent', protein_pct: 0.15, fiber_per_kcal: 0.014 },
};

interface LifeStageCfg {
  label: string;
  protein_mode: ProteinMode;
  protein_gkg?: number;
  protein_add_g?: number;
  kcal_add: number;
}

const LIFE_STAGES: Record<string, LifeStageCfg> = {
  default: { label: 'Обычный взрослый', protein_mode: null, kcal_add: 0 },
  athlete_endurance: { label: 'Спортсмен на выносливость', protein_mode: 'gkg', protein_gkg: 1.3, kcal_add: 0 },
  athlete_strength: { label: 'Спортсмен-силовик', protein_mode: 'gkg', protein_gkg: 1.8, kcal_add: 0 },
  older_adult: { label: 'Пожилой (60+)', protein_mode: 'gkg', protein_gkg: 1.1, kcal_add: 0 },
  pregnant: { label: 'Беременность (2-3 триместр)', protein_mode: 'add_g', protein_add_g: 25, kcal_add: 340 },
  lactating: { label: 'Кормление грудью', protein_mode: 'add_g', protein_add_g: 25, kcal_add: 500 },
};

// --- Типы -------------------------------------------------------------------

export interface CalcInput {
  sex: 'male' | 'female' | 'other' | null;
  age: number;
  weight: number; // кг — берётся ПОСЛЕДНИЙ вес из weight_log
  height: number; // см
  activity: string | null;
  goal: string | null;
  condition?: string | null;
  life_stage?: string | null;
  formula?: 'who' | 'mifflin';
}

export interface CalcResult {
  target_kcal: number;
  protein_g: number;
  fat_g: number;
  carbs_g: number;
  fiber_g: number;
  bmr: number;
  tdee: number;
  condition_label: string;
  activity_label: string;
  goal_label: string;
  warnings: string[];
}

// --- Формулы BMR ------------------------------------------------------------

function calcBmrMifflin(sex: string, weight: number, height: number, age: number): number {
  const base = 10 * weight + 6.25 * height - 5 * age;
  return sex === 'male' ? base + 5 : base - 161;
}

// Уравнения Скофилда (ФАО/ВОЗ/ООН). W=кг, H=метры.
function whoEquation(sex: string, age: number, W: number, Hm: number): number {
  if (sex === 'male') {
    if (age >= 10 && age < 18) return 16.6 * W + 77.0 * Hm + 572.0;
    if (age >= 18 && age < 30) return 15.4 * W - 27.0 * Hm + 717.0;
    if (age >= 30 && age < 60) return 11.3 * W + 16.0 * Hm + 901.0;
    if (age >= 60) return 8.8 * W + 1128.0 * Hm - 1071.0;
  } else {
    // female / other считаем по женской формуле (мягкое допущение для 'other')
    if (age >= 10 && age < 18) return 7.4 * W + 482.0 * Hm + 217.0;
    if (age >= 18 && age < 30) return 13.3 * W + 334.0 * Hm + 35.0;
    if (age >= 30 && age < 60) return 8.7 * W - 25.0 * Hm + 865.0;
    if (age >= 60) return 9.2 * W + 637.0 * Hm - 302.0;
  }
  // возраст вне табличных диапазонов (< 10) — калькулятор не для детей,
  // возвращаем mifflin как разумный fallback
  return calcBmrMifflin(sex, W, Hm * 100, age);
}

function calcBmrWho(sex: string, age: number, weight: number, heightCm: number): number {
  return whoEquation(sex, age, weight, heightCm / 100.0);
}

// --- Белок (приоритет: CKD > стадия жизни > condition) ----------------------

function proteinFromCondition(weight: number, cond: ConditionCfg, kcalBasis: number): number {
  if (cond.protein_mode === 'gkg') return (cond.protein_gkg ?? 0) * weight;
  return (kcalBasis * (cond.protein_pct ?? 0)) / KCAL_PER_GRAM.protein;
}

function calcProtein(
  weight: number, conditionKey: string, lifeStageKey: string, kcalBasis: number,
): { protein: number; warning: string | null } {
  const cond = CONDITIONS[conditionKey] ?? CONDITIONS.healthy;
  const stage = LIFE_STAGES[lifeStageKey] ?? LIFE_STAGES.default;

  // 1) CKD имеет абсолютный приоритет
  if (conditionKey === 'ckd') {
    let warning: string | null = null;
    if (stage.protein_mode === 'gkg' || stage.protein_mode === 'add_g') {
      warning = `При ХБП белок ограничен до ${cond.protein_gkg} г/кг вне зависимости от стадии жизни (${stage.label}). Стадию/диализ уточняет нефролог.`;
    }
    return { protein: (cond.protein_gkg ?? 0.8) * weight, warning };
  }

  // 2) Стадия жизни
  if (stage.protein_mode === 'gkg') {
    return { protein: (stage.protein_gkg ?? 0) * weight, warning: null };
  }
  if (stage.protein_mode === 'add_g') {
    const base = proteinFromCondition(weight, cond, kcalBasis);
    return { protein: base + (stage.protein_add_g ?? 0), warning: null };
  }

  // 3) Condition
  return { protein: proteinFromCondition(weight, cond, kcalBasis), warning: null };
}

// --- Главный расчёт ---------------------------------------------------------

export function calculate(input: CalcInput): CalcResult {
  const {
    sex, age, weight, height, activity, goal,
    condition, life_stage, formula = 'who',
  } = input;

  const sexKey = (sex ?? 'female') as 'male' | 'female';
  const activityKey = activity ?? 'sedentary';
  const goalKey = goal ?? 'maintain';

  // Маппинг ключей онбординга → калькулятор
  const condKey = CONDITION_MAP[condition ?? 'healthy'] ?? 'healthy';
  const stageKey = life_stage ?? 'default';
  const actCfg = ACTIVITY_LEVELS[activityKey] ?? ACTIVITY_LEVELS.sedentary;
  const goalCfg = GOALS[goalKey] ?? GOALS.maintain;

  // BMR
  const bmrMifflin = calcBmrMifflin(sexKey, weight, height, age);
  const bmrWho = calcBmrWho(sexKey, age, weight, height);
  const bmr = formula === 'mifflin' ? bmrMifflin : bmrWho;

  // TDEE
  const tdee = bmr * actCfg.factor;
  const tdeeMifflin = bmrMifflin * actCfg.factor;
  const tdeeWho = bmrWho * actCfg.factor;
  const tdeeUsed = formula === 'mifflin' ? tdeeMifflin : tdeeWho;

  // Целевые ккал = TDEE + Δ цели + прибавка стадии, не ниже минимума
  const stage = LIFE_STAGES[stageKey] ?? LIFE_STAGES.default;
  const minKcal = MIN_KCAL[sexKey] ?? 1200;
  const targetKcal = Math.max(tdeeUsed + goalCfg.kcal_delta + stage.kcal_add, minKcal);

  // БЖУ
  const { protein: proteinG, warning } = calcProtein(weight, condKey, stageKey, targetKcal);
  const proteinKcal = proteinG * KCAL_PER_GRAM.protein;
  const fatG = (targetKcal * FAT_SHARE) / KCAL_PER_GRAM.fat;
  const fatKcal = fatG * KCAL_PER_GRAM.fat;
  const carbsKcal = targetKcal - proteinKcal - fatKcal;
  const carbsG = carbsKcal / KCAL_PER_GRAM.carbs;

  // Клетчатка
  const cond = CONDITIONS[condKey] ?? CONDITIONS.healthy;
  const fiberG = Math.max(targetKcal * cond.fiber_per_kcal, MIN_FIBER_G);

  // Предупреждения
  const warnings: string[] = [];
  if (warning) warnings.push(warning);
  if (stageKey === 'pregnant' && goalKey === 'lose') {
    warnings.push('Похудение во время беременности не рекомендуется — нужна консультация врача.');
  }

  return {
    target_kcal: targetKcal,
    protein_g: proteinG,
    fat_g: fatG,
    carbs_g: carbsG,
    fiber_g: fiberG,
    bmr,
    tdee: tdeeUsed,
    condition_label: cond.label,
    activity_label: actCfg.label,
    goal_label: goalCfg.label,
    warnings,
  };
}
